File: src/python/underchex/moves.py
"""
Underchex Move Generation and Validation

Signed-by: agent #17 claude-sonnet-4 via opencode 20260122T05:47:09
Edited-by: agent #19 claude-sonnet-4 via amp 20260122T06:10:50 (fix: Python 3.9 compatibility)
"""
from __future__ import annotations

from .types import (
    HexCoord,
    Direction,
    ALL_DIRECTIONS,
    DIAGONAL_DIRECTIONS,
    LANCE_A_DIRECTIONS,
    LANCE_B_DIRECTIONS,
    PieceType,
    Piece,
    Color,
    BoardState,
    Move,
    MoveValidation,
    coord_to_string,
    coords_equal,
    opposite_color,
    is_promotion_zone,
    PROMOTION_TARGETS,
)

from .board import (
    is_valid_cell,
    get_neighbor,
    get_ray,
    get_knight_targets,
)

# ============================================================================
# Piece Movement Patterns
# ============================================================================

def get_forward_direction(color: Color) -> Direction:
    """Get the forward direction for a color."""
    return "N" if color == "white" else "S"


def get_pawn_capture_directions(color: Color) -> tuple[Direction, ...]:
    """Get pawn capture directions for a color."""
    return ("N", "NE", "NW") if color == "white" else ("S", "SE", "SW")


def get_piece_directions(piece: Piece) -> tuple[Direction, ...]:
    """Get directions a piece can move in."""
    if piece.type in ("king", "queen"):
        return ALL_DIRECTIONS
    elif piece.type == "chariot":
        return DIAGONAL_DIRECTIONS
    elif piece.type == "lance":
        return LANCE_A_DIRECTIONS if piece.variant == "A" else LANCE_B_DIRECTIONS
    else:
        return ()


def is_slider(piece_type: PieceType) -> bool:
    """Check if a piece is a slider (can move multiple squares)."""
    return piece_type in ("queen", "lance", "chariot")


# ============================================================================
# Move Generation
# ============================================================================

def get_piece_at(board: BoardState, coord: HexCoord) -> Piece | None:
    """Get piece at a position, or None if empty."""
    return board.get(coord_to_string(coord))


def is_occupied(board: BoardState, coord: HexCoord) -> bool:
    """Check if a cell is occupied."""
    return coord_to_string(coord) in board


def has_enemy(board: BoardState, coord: HexCoord, color: Color) -> bool:
    """Check if a cell has an enemy piece."""
    piece = get_piece_at(board, coord)
    return piece is not None and piece.color != color


def has_friendly(board: BoardState, coord: HexCoord, color: Color) -> bool:
    """Check if a cell has a friendly piece."""
    piece = get_piece_at(board, coord)
    return piece is not None and piece.color == color


def generate_pseudo_legal_moves(
    board: BoardState,
    piece: Piece,
    from_coord: HexCoord
) -> list[Move]:
    """Generate pseudo-legal moves for a piece (doesn't check for leaving king in check)."""
    moves: list[Move] = []
    
    if piece.type == "pawn":
        _generate_pawn_moves(board, piece, from_coord, moves)
    elif piece.type == "king":
        _generate_king_moves(board, piece, from_coord, moves)
    elif piece.type == "knight":
        _generate_knight_moves(board, piece, from_coord, moves)
    elif piece.type in ("queen", "lance", "chariot"):
        _generate_slider_moves(board, piece, from_coord, moves)
    
    return moves


def _generate_pawn_moves(
    board: BoardState,
    piece: Piece,
    from_coord: HexCoord,
    moves: list[Move]
) -> None:
    """Generate pawn moves."""
    forward_dir = get_forward_direction(piece.color)
    capture_directions = get_pawn_capture_directions(piece.color)
    
    # Forward move (non-capture)
    forward = get_neighbor(from_coord, forward_dir)
    if forward is not None and not is_occupied(board, forward):
        if is_promotion_zone(forward, piece.color):
            # Generate promotion moves
            for promotion_type in PROMOTION_TARGETS:
                moves.append(_create_move(piece, from_coord, forward, None, promotion_type))
        else:
            moves.append(_create_move(piece, from_coord, forward))
    
    # Captures (including forward capture)
    for direction in capture_directions:
        target = get_neighbor(from_coord, direction)
        if target is not None and has_enemy(board, target, piece.color):
            captured = get_piece_at(board, target)
            if is_promotion_zone(target, piece.color):
                # Generate promotion captures
                for promotion_type in PROMOTION_TARGETS:
                    moves.append(_create_move(piece, from_coord, target, captured, promotion_type))
            else:
                moves.append(_create_move(piece, from_coord, target, captured))


def _generate_king_moves(
    board: BoardState,
    piece: Piece,
    from_coord: HexCoord,
    moves: list[Move]
) -> None:
    """Generate king moves."""
    for direction in ALL_DIRECTIONS:
        target = get_neighbor(from_coord, direction)
        if target is not None and not has_friendly(board, target, piece.color):
            captured = get_piece_at(board, target)
            moves.append(_create_move(piece, from_coord, target, captured))


def _generate_knight_moves(
    board: BoardState,
    piece: Piece,
    from_coord: HexCoord,
    moves: list[Move]
) -> None:
    """Generate knight moves."""
    for target in get_knight_targets(from_coord):
        if not has_friendly(board, target, piece.color):
            captured = get_piece_at(board, target)
            moves.append(_create_move(piece, from_coord, target, captured))


def _generate_slider_moves(
    board: BoardState,
    piece: Piece,
    from_coord: HexCoord,
    moves: list[Move]
) -> None:
    """Generate slider (queen, lance, chariot) moves."""
    directions = get_piece_directions(piece)
    
    for direction in directions:
        ray = get_ray(from_coord, direction)
        for target in ray:
            if has_friendly(board, target, piece.color):
                break  # Blocked by friendly piece
            captured = get_piece_at(board, target)
            moves.append(_create_move(piece, from_coord, target, captured))
            if captured is not None:
                break  # Can't move past a captured piece


def _create_move(
    piece: Piece,
    from_coord: HexCoord,
    to_coord: HexCoord,
    captured: Piece | None = None,
    promotion: PieceType | None = None
) -> Move:
    """Create a Move object."""
    return Move(
        from_coord=from_coord,
        to_coord=to_coord,
        piece=piece,
        captured=captured,
        promotion=promotion,
    )


# ============================================================================
# Check Detection
# ============================================================================

def find_king(board: BoardState, color: Color) -> HexCoord | None:
    """Find the king of a given color."""
    for pos_str, piece in board.items():
        if piece.type == "king" and piece.color == color:
            parts = pos_str.split(",")
            q = int(parts[0]) if len(parts) > 0 else 0
            r = int(parts[1]) if len(parts) > 1 else 0
            return HexCoord(q=q, r=r)
    return None


def is_attacked(
    board: BoardState,
    target: HexCoord,
    by_color: Color
) -> bool:
    """Check if a square is attacked by any piece of the given color."""
    # Check for pawn attacks
    pawn_capture_directions = get_pawn_capture_directions(by_color)
    for direction in pawn_capture_directions:
        reverse_dir = _get_opposite_direction(direction)
        attacker = get_neighbor(target, reverse_dir)
        if attacker is not None:
            piece = get_piece_at(board, attacker)
            if piece is not None and piece.type == "pawn" and piece.color == by_color:
                return True
    
    # Check for king attacks
    for direction in ALL_DIRECTIONS:
        attacker = get_neighbor(target, direction)
        if attacker is not None:
            piece = get_piece_at(board, attacker)
            if piece is not None and piece.type == "king" and piece.color == by_color:
                return True
    
    # Check for knight attacks
    for attacker_pos in get_knight_targets(target):
        piece = get_piece_at(board, attacker_pos)
        if piece is not None and piece.type == "knight" and piece.color == by_color:
            return True
    
    # Check for slider attacks (queen, lance, chariot)
    for direction in ALL_DIRECTIONS:
        ray = get_ray(target, direction)
        for pos in ray:
            piece = get_piece_at(board, pos)
            if piece is None:
                continue
            if piece.color != by_color:
                break
            
            # Check if this piece can attack along this direction
            piece_directions = get_piece_directions(piece)
            reverse_dir = _get_opposite_direction(direction)
            if reverse_dir in piece_directions and is_slider(piece.type):
                return True
            break  # Blocked by this piece either way
    
    return False


def _get_opposite_direction(direction: Direction) -> Direction:
    """Get the opposite direction."""
    opposites: dict[Direction, Direction] = {
        "N": "S", "S": "N",
        "NE": "SW", "SW": "NE",
        "NW": "SE", "SE": "NW",
    }
    return opposites[direction]


def is_in_check(board: BoardState, color: Color) -> bool:
    """Check if the king of a given color is in check."""
    king_pos = find_king(board, color)
    if king_pos is None:
        return False  # No king - shouldn't happen in valid game
    return is_attacked(board, king_pos, opposite_color(color))


# ============================================================================
# Legal Move Generation
# ============================================================================

def apply_move(board: BoardState, move: Move) -> BoardState:
    """
    Apply a move to a board state (returns new board state).
    Handles pawn promotion by replacing the piece.
    """
    new_board = dict(board)
    del new_board[coord_to_string(move.from_coord)]
    
    # Handle promotion
    if move.promotion is not None:
        promoted_piece = Piece(
            type=move.promotion,
            color=move.piece.color,
            variant="A" if move.promotion == "lance" else None,
        )
        new_board[coord_to_string(move.to_coord)] = promoted_piece
    else:
        new_board[coord_to_string(move.to_coord)] = move.piece
    
    return new_board


def generate_legal_moves(
    board: BoardState,
    piece: Piece,
    from_coord: HexCoord
) -> list[Move]:
    """Generate all legal moves for a piece."""
    pseudo_legal = generate_pseudo_legal_moves(board, piece, from_coord)
    
    return [
        move for move in pseudo_legal
        if not is_in_check(apply_move(board, move), piece.color)
    ]


def generate_all_legal_moves(board: BoardState, color: Color) -> list[Move]:
    """Generate all legal moves for a player."""
    moves: list[Move] = []
    
    for pos_str, piece in board.items():
        if piece.color != color:
            continue
        parts = pos_str.split(",")
        q = int(parts[0]) if len(parts) > 0 else 0
        r = int(parts[1]) if len(parts) > 1 else 0
        from_coord = HexCoord(q=q, r=r)
        moves.extend(generate_legal_moves(board, piece, from_coord))
    
    return moves


# ============================================================================
# Move Validation
# ============================================================================

def validate_move(
    board: BoardState,
    from_coord: HexCoord,
    to_coord: HexCoord,
    turn: Color
) -> MoveValidation:
    """Validate a specific move."""
    piece = get_piece_at(board, from_coord)
    
    if piece is None:
        return MoveValidation(legal=False, reason="noPieceAtSource")
    
    if piece.color != turn:
        return MoveValidation(legal=False, reason="notYourPiece")
    
    if not is_valid_cell(to_coord):
        return MoveValidation(legal=False, reason="invalidDestination")
    
    legal_moves = generate_legal_moves(board, piece, from_coord)
    matching_move = next(
        (m for m in legal_moves if coords_equal(m.to_coord, to_coord)),
        None
    )
    
    if matching_move is None:
        # Check if it would leave king in check
        pseudo_legal = generate_pseudo_legal_moves(board, piece, from_coord)
        pseudo_match = next(
            (m for m in pseudo_legal if coords_equal(m.to_coord, to_coord)),
            None
        )
        
        if pseudo_match is not None:
            return MoveValidation(legal=False, reason="movesIntoCheck")
        
        return MoveValidation(legal=False, reason="illegalMove")
    
    return MoveValidation(
        legal=True,
        capture=matching_move.captured is not None
    )

File: src/elixir/lib/underchex/moves.ex
defmodule Underchex.Moves do
  @moduledoc """
  Move generation and validation for Underchex.

  Signed-by: agent #26 claude-sonnet-4 via opencode 20260122T07:33:49
  """

  alias Underchex.{Types, Board}

  @doc """
  Generate pseudo-legal moves for a piece (doesn't check for leaving king in check).
  """
  @spec generate_pseudo_legal_moves(Types.board(), Types.piece(), Types.coord()) :: [Types.move()]
  def generate_pseudo_legal_moves(board, piece, from) do
    case piece.type do
      :pawn -> generate_pawn_moves(board, piece, from)
      :king -> generate_king_moves(board, piece, from)
      :knight -> generate_knight_moves(board, piece, from)
      type when type in [:queen, :lance, :chariot] -> generate_slider_moves(board, piece, from)
      _ -> []
    end
  end

  defp generate_pawn_moves(board, piece, from) do
    forward_dir = Types.forward_direction(piece.color)
    capture_dirs = Types.pawn_capture_directions(piece.color)

    # Forward move (non-capture)
    forward_moves =
      case Board.get_neighbor(from, forward_dir) do
        nil ->
          []

        forward ->
          if Board.occupied?(board, forward) do
            []
          else
            if Types.promotion_zone?(forward, piece.color) do
              # Generate all promotion moves
              Types.promotion_targets()
              |> Enum.map(&Types.new_move(piece, from, forward, nil, &1))
            else
              [Types.new_move(piece, from, forward)]
            end
          end
      end

    # Captures
    capture_moves =
      capture_dirs
      |> Enum.flat_map(fn dir ->
        case Board.get_neighbor(from, dir) do
          nil ->
            []

          target ->
            if Board.has_enemy?(board, target, piece.color) do
              captured = Board.get_piece_at(board, target)

              if Types.promotion_zone?(target, piece.color) do
                Types.promotion_targets()
                |> Enum.map(&Types.new_move(piece, from, target, captured, &1))
              else
                [Types.new_move(piece, from, target, captured)]
              end
            else
              []
            end
        end
      end)

    forward_moves ++ capture_moves
  end

  defp generate_king_moves(board, piece, from) do
    Types.all_directions()
    |> Enum.flat_map(fn dir ->
      case Board.get_neighbor(from, dir) do
        nil ->
          []

        target ->
          if Board.has_friendly?(board, target, piece.color) do
            []
          else
            captured = Board.get_piece_at(board, target)
            [Types.new_move(piece, from, target, captured)]
          end
      end
    end)
  end

  defp generate_knight_moves(board, piece, from) do
    Board.get_knight_targets(from)
    |> Enum.flat_map(fn target ->
      if Board.has_friendly?(board, target, piece.color) do
        []
      else
        captured = Board.get_piece_at(board, target)
        [Types.new_move(piece, from, target, captured)]
      end
    end)
  end

  defp generate_slider_moves(board, piece, from) do
    directions = Types.piece_directions(piece)

    directions
    |> Enum.flat_map(fn dir ->
      Board.get_ray(from, dir)
      |> Enum.reduce_while([], fn target, acc ->
        cond do
          Board.has_friendly?(board, target, piece.color) ->
            {:halt, acc}

          Board.has_enemy?(board, target, piece.color) ->
            captured = Board.get_piece_at(board, target)
            {:halt, [Types.new_move(piece, from, target, captured) | acc]}

          true ->
            {:cont, [Types.new_move(piece, from, target) | acc]}
        end
      end)
    end)
  end

  @doc """
  Apply a move to a board state.
  """
  @spec apply_move(Types.board(), Types.move()) :: Types.board()
  def apply_move(board, move) do
    board
    |> Map.delete(move.from)
    |> Map.put(move.to, promoted_piece(move))
  end

  defp promoted_piece(move) do
    case move.promotion do
      nil ->
        move.piece

      promotion_type ->
        # Default lance variant to :a for promotions
        variant = if promotion_type == :lance, do: :a, else: nil
        Types.new_piece(promotion_type, move.piece.color, variant)
    end
  end

  @doc """
  Check if a square is attacked by any piece of the given color.
  """
  @spec attacked?(Types.board(), Types.coord(), Types.color()) :: boolean()
  def attacked?(board, target, by_color) do
    attacked_by_pawn?(board, target, by_color) or
      attacked_by_king?(board, target, by_color) or
      attacked_by_knight?(board, target, by_color) or
      attacked_by_slider?(board, target, by_color)
  end

  defp attacked_by_pawn?(board, target, by_color) do
    capture_dirs = Types.pawn_capture_directions(by_color)

    Enum.any?(capture_dirs, fn dir ->
      reverse_dir = Types.opposite_direction(dir)

      case Board.get_neighbor(target, reverse_dir) do
        nil ->
          false

        attacker_pos ->
          case Board.get_piece_at(board, attacker_pos) do
            %{type: :pawn, color: ^by_color} -> true
            _ -> false
          end
      end
    end)
  end

  defp attacked_by_king?(board, target, by_color) do
    Enum.any?(Types.all_directions(), fn dir ->
      case Board.get_neighbor(target, dir) do
        nil ->
          false

        attacker_pos ->
          case Board.get_piece_at(board, attacker_pos) do
            %{type: :king, color: ^by_color} -> true
            _ -> false
          end
      end
    end)
  end

  defp attacked_by_knight?(board, target, by_color) do
    Board.get_knight_targets(target)
    |> Enum.any?(fn attacker_pos ->
      case Board.get_piece_at(board, attacker_pos) do
        %{type: :knight, color: ^by_color} -> true
        _ -> false
      end
    end)
  end

  defp attacked_by_slider?(board, target, by_color) do
    Enum.any?(Types.all_directions(), fn dir ->
      Board.get_ray(target, dir)
      |> Enum.reduce_while(false, fn pos, _acc ->
        case Board.get_piece_at(board, pos) do
          nil ->
            {:cont, false}

          %{color: color} when color != by_color ->
            {:halt, false}

          piece ->
            reverse_dir = Types.opposite_direction(dir)
            piece_dirs = Types.piece_directions(piece)

            if reverse_dir in piece_dirs and Types.slider?(piece.type) do
              {:halt, true}
            else
              {:halt, false}
            end
        end
      end)
    end)
  end

  @doc """
  Check if the king of a given color is in check.
  """
  @spec in_check?(Types.board(), Types.color()) :: boolean()
  def in_check?(board, color) do
    case Board.find_king(board, color) do
      nil -> false
      king_pos -> attacked?(board, king_pos, Types.opposite_color(color))
    end
  end

  @doc """
  Generate all legal moves for a piece.
  """
  @spec generate_legal_moves(Types.board(), Types.piece(), Types.coord()) :: [Types.move()]
  def generate_legal_moves(board, piece, from) do
    generate_pseudo_legal_moves(board, piece, from)
    |> Enum.filter(fn move ->
      new_board = apply_move(board, move)
      not in_check?(new_board, piece.color)
    end)
  end

  @doc """
  Generate all legal moves for a player.
  """
  @spec generate_all_legal_moves(Types.board(), Types.color()) :: [Types.move()]
  def generate_all_legal_moves(board, color) do
    board
    |> Enum.flat_map(fn {coord, piece} ->
      if piece.color == color do
        generate_legal_moves(board, piece, coord)
      else
        []
      end
    end)
  end

  @doc """
  Validate a move.
  """
  @spec validate_move(Types.board(), Types.coord(), Types.coord(), Types.color()) ::
          {:ok, Types.move()} | {:error, atom()}
  def validate_move(board, from, to, turn) do
    case Board.get_piece_at(board, from) do
      nil ->
        {:error, :no_piece_at_source}

      piece when piece.color != turn ->
        {:error, :not_your_piece}

      piece ->
        if not Types.valid_cell?(to) do
          {:error, :invalid_destination}
        else
          legal_moves = generate_legal_moves(board, piece, from)

          case Enum.find(legal_moves, fn m -> m.to == to end) do
            nil ->
              # Check if it would leave king in check
              pseudo = generate_pseudo_legal_moves(board, piece, from)

              if Enum.any?(pseudo, fn m -> m.to == to end) do
                {:error, :moves_into_check}
              else
                {:error, :illegal_move}
              end

            move ->
              {:ok, move}
          end
        end
    end
  end
end
